clash-lib/src/Clash/Rewrite/Util.hs

```diff
@@ -1136,6 +1183,13 @@ bindPureHeap tcm heap rw (TransformContext is0 hist) e = do
         ty = termType tcm term
         nm = mkLocalId ty (mkUnsafeSystemName "x" uniq) -- See [Note: Name re-creation]
 
+    inlineTest _ (i, stripTicks -> e_) =
+      if isLocalVar e_ then
+        -- Don't inline `let x = x in x`, it throws  us in an infinite loop
```

Comment:
inlineBinders already does this check: https://github.com/clash-lang/clash-compiler/blob/3762801bcfe4be851639fadbc46170a493f3bee4/clash-lib/src/Clash/Rewrite/Util.hs#L474